File: 5-Problem-A-Day/77. Combinations.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> ans = new LinkedList<>();
        backtrack(n, 1, k, new LinkedList<>(), ans);
        return ans;
    }

    private void backtrack(int n, int start, int k, List<Integer> curr, List<List<Integer>> ans) {
        if (curr.size() == k)
            ans.add(new LinkedList<>(curr));
        else {
            for (int i = start; i <= n; i++) {
                curr.add(i);
                backtrack(n, i + 1, k, curr, ans);
                curr.remove(curr.size() - 1);
            }
        }
    }
}

File: 2024/77. Combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<int> curr;
        vector<vector<int>> res;
        backtrack(1, n, k, curr, res);
        return res;
    }

    void backtrack(int start, int n, int k, vector<int>& curr, vector<vector<int>>& res) {
        if (k == 0) {
            res.push_back(curr);
            return;
        }
        for (int i = start; i <= n; i++) {
            curr.push_back(i);
            backtrack(i + 1, n, k - 1, curr, res);
            curr.pop_back();
        }
    }
};